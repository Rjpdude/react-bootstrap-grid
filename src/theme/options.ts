export interface ThemeOptions {
  /**
   * The `margin` and `padding` in pixels for all grid elements.
   *
   * @default 15
   */
  spacing: number
  /**
   * The maximum width in pixels of the small breakpoint container.
   *
   * @default 576
   */
  sm: number
  /**
   * The width in pixels to assign to the small breakpoint container.
   *
   * @default 540
   */
  smWidth: number
  /**
   * The maximum width in pixels of the medium breakpoint container.
   *
   * @default 768
   */
  md: number
  /**
   * The width in pixels to assign to the medium breakpoint container.
   *
   * @default 720
   */
  mdWidth: number
  /**
   * The maximum width in pixels of the large breakpoint container.
   *
   * @default 992
   */
  lg: number
  /**
   * The width in pixels to assign to the large breakpoint container.
   *
   * @default 960
   */
  lgWidth: number
  /**
   * The maximum width in pixels of the extra-large breakpoint container.
   *
   * @default 1200
   */
  xl: number
  /**
   * The width in pixels to assign to the extra-large breakpoint container.
   *
   * @default 1140
   */
  xlWidth: number
}

export function getDefaultTheme(): ThemeOptions {
  return {
    spacing: 15,
    sm: 576,
    smWidth: 540,
    md: 768,
    mdWidth: 720,
    lg: 992,
    lgWidth: 960,
    xl: 1200,
    xlWidth: 1140
  }
}
