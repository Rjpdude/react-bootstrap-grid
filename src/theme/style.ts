import { css, createGlobalStyle } from 'styled-components'

export const Global = css`
  html {
    box-sizing: border-box;
    -ms-overflow-style: scrollbar;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }
`

export const GridThemeStyle = createGlobalStyle`
  ${Global}
`
