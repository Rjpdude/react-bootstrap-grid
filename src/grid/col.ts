import { CSSObject } from 'styled-components'
import { ThemeOptions } from '../theme/options'
import { GridElement } from '../element/interface'
import { configureElement } from '../element/element'
import { getElementPropertyStyle } from '../element/util'
import * as gridPropertyStyleMap from '../element/style'

export interface GridColumn {
  /**
   * The size to assume on this breakpoint. If undefined, the breakpoint will
   * assume the size of the next lowest supplied size, or equal.
   *
   * @see https://getbootstrap.com/docs/4.0/layout/grid/#auto-layout-columns
   * @default equal
   */
  size: BreakpointVal
}

export interface ColumnProps {
  /**
   * The column's extra small / default size.
   *
   * @see https://getbootstrap.com/docs/4.0/layout/grid/#auto-layout-columns
   * @default equal
   */
  size: BreakpointVal
  /**
   * Either the size of the small breakpoint, or an object describing the
   * breakpoint's size and/or styling properties.
   */
  sm: BreakpointPropVal
  /**
   * Either the size of the medium breakpoint, or an object describing the
   * breakpoint's size and/or styling properties.
   */
  md: BreakpointPropVal
  /**
   * Either the size of the large breakpoint, or an object describing the
   * breakpoint's size and/or styling properties.
   */
  lg: BreakpointPropVal
  /**
   * Either the size of the extra-large breakpoint, or an object describing the
   * breakpoint's size and/or styling properties.
   */
  xl: BreakpointPropVal
}

export type BreakpointSize = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
export type BreakpointVal = 'equal' | 'auto' | string | BreakpointSize
export type BreakpointPropVal = BreakpointVal | GridElement<GridColumn>

export const Col = configureElement<ColumnProps>(
  ({ ownProps, themeProps }) => {
    if (!ownProps.size) {
      ownProps.size = 'equal'
    }

    return Object.keys(ownProps).reduce(
      (style, breakpointId) =>
        assignBreakpointStyle(
          ownProps,
          themeProps,
          breakpointId as keyof ColumnProps,
          style
        ),
      initialColStyle(themeProps)
    )
  },
  ['size', 'sm', 'md', 'lg', 'xl']
)

export function assignBreakpointStyle(
  props: Partial<ColumnProps>,
  theme: ThemeOptions,
  breakpointId: keyof ColumnProps,
  style: CSSObject
) {
  const breakpoint = props[breakpointId]
  const size = typeof breakpoint === 'object' ? breakpoint.size : breakpoint
  const baseStyle = generateBreakpointStyle(breakpointId, size)

  if (typeof breakpoint === 'object') {
    Object.keys(breakpoint).forEach((key) => {
      Object.assign(
        baseStyle,
        getElementPropertyStyle(gridPropertyStyleMap, key, breakpoint[key])
      )
    })
  }

  return Object.assign(
    style,
    breakpointId === 'size'
      ? baseStyle
      : {
          [`@media (min-width: ${theme[breakpointId]}px)`]: baseStyle
        }
  )
}

export function generateBreakpointStyle(
  breakpointId: keyof ColumnProps,
  size: BreakpointVal
) {
  return !size || size === 'equal'
    ? !size && breakpointId !== 'size'
      ? {}
      : breakpointStyleEqual
    : size === 'auto'
    ? breakpointStyleAuto
    : breakpointStyle(size)
}

export function initialColStyle({ spacing }: ThemeOptions): CSSObject {
  const padding = `${spacing}px`
  return {
    width: '100%',
    position: 'relative',
    paddingRight: padding,
    paddingLeft: padding
  }
}

export function breakpointStyle(breakpointVal: string | number): CSSObject {
  const size =
    typeof breakpointVal === 'string'
      ? breakpointVal
      : `${((100 / 12) * breakpointVal).toFixed(6)}%`
  return {
    flex: `0 0 ${size}`,
    maxWidth: `${size}`
  }
}

export const breakpointStyleAuto: CSSObject = {
  width: 'auto',
  maxWidth: '100%',
  flex: '0 0 auto'
}

export const breakpointStyleEqual: CSSObject = {
  maxWidth: '100%',
  flexBasis: 0,
  flexGrow: 1
}
