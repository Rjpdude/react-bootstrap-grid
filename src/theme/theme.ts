import { createElement, createContext, PropsWithChildren } from 'react'
import { ThemeOptions, getDefaultTheme } from './options'
import { GridThemeStyle } from './style'

export const gridThemeContext = createContext(getDefaultTheme())

export function GridTheme(props: PropsWithChildren<Partial<ThemeOptions>>) {
  const style = createElement(GridThemeStyle)
  const context = {
    value: getMergedTheme(props)
  }
  return createElement(
    gridThemeContext.Provider,
    context,
    props.themeStyle !== false && style,
    props.children
  )
}

export function getMergedTheme(themeProps: Partial<ThemeOptions>) {
  const theme = getDefaultTheme()

  Object.keys(theme).forEach((key) => {
    if (themeProps[key]) {
      Object.assign(theme, {
        [key]: themeProps[key]
      })
    }
  })

  return theme
}
