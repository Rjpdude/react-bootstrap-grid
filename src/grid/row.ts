import { cloneElement, ReactElement } from 'react'
import { CSSObject } from 'styled-components'
import { configureElement } from '../element/element'
import { GridElement } from '../element/interface'
import { ThemeOptions } from '../theme/options'
import { Col } from './col'

export interface RowProps {
  /**
   * When false, the row's margin along with the padding of child columns are disabled.
   *
   * @see https://getbootstrap.com/docs/4.4/layout/grid/#no-gutters
   * @default true
   */
  gutters: boolean
  /**
   * The number of columns (between 1 and 6) to render per row.
   *
   * @see https://getbootstrap.com/docs/4.4/layout/grid/#row-columns
   */
  cols: RowColsVal
  /**
   * The number of columns (between 1 and 6) to render per row at the `small` breakpoint.
   *
   * @see https://getbootstrap.com/docs/4.4/layout/grid/#row-columns
   */
  cols_sm: RowColsVal
  /**
   * The number of columns (between 1 and 6) to render per row at the `medium` breakpoint.
   *
   * @see https://getbootstrap.com/docs/4.4/layout/grid/#row-columns
   */
  cols_md: RowColsVal
  /**
   * The number of columns (between 1 and 6) to render per row at the `large` breakpoint.
   *
   * @see https://getbootstrap.com/docs/4.4/layout/grid/#row-columns
   */
  cols_lg: RowColsVal
  /**
   * The number of columns (between 1 and 6) to render per row at the `extra large` breakpoint.
   *
   * @see https://getbootstrap.com/docs/4.4/layout/grid/#row-columns
   */
  cols_xl: RowColsVal
  /**
   * The element properties to assign to child columns.
   */
  col_props: RowColProps
  /**
   * The element properties to assign to child columns at the `small` breakpoint.
   */
  col_props_sm: RowColProps
  /**
   * The element properties to assign to child columns at the `medium` breakpoint.
   */
  col_props_md: RowColProps
  /**
   * The element properties to assign to child columns at the `large` breakpoint.
   */
  col_props_lg: RowColProps
  /**
   * The element properties to assign to child columns at the `extra large` breakpoint.
   */
  col_props_xl: RowColProps
}

export type RowColsVal = 1 | 2 | 3 | 4 | 5 | 6
export type RowColProps = Omit<GridElement<{}>, 'order'>

export const rowPropKeys = [
  'gutters',
  'cols',
  'cols_sm',
  'cols_md',
  'cols_lg',
  'cols_xl',
  'col_props',
  'col_props_sm',
  'col_props_md',
  'col_props_lg',
  'col_props_xl'
]

export const Row = configureElement<RowProps>(
  'twbs-grid-row', ({ ownProps, themeProps }) => {
    const rowStyle = initialRowStyle(themeProps, ownProps.gutters === false)

    if (ownProps.gutters === false) {
      rowStyle[`> ${Col.styledComponent}`] = noGutterStyle
    }

    return rowStyle
  },
  rowPropKeys,
  rowChildMapper
)

export function rowChildMapper(child: ReactElement<any, any>, props: RowProps) {
  return !child?.type?.styledComponent
    ? child
    : cloneElement(
        child,
        Object.keys(props).reduce(
          (childProps, propKey) => {
            return propKey.startsWith('cols')
              ? assignColSize(propKey, props[propKey], childProps)
              : propKey.startsWith('col_props')
              ? assignColProp(propKey, props[propKey], childProps)
              : childProps
          },
          { ...child.props }
        )
      )
}

export function assignColSize(key: string, val: number, columnProps: any) {
  const propKey = key === 'cols' ? 'size' : key.substr(5)
  const size = val === 5 ? '20%' : 12 / val

  if (!columnProps[propKey]) {
    columnProps[propKey] = size
  } else if (
    typeof columnProps[propKey] === 'object' &&
    !columnProps[propKey].size
  ) {
    columnProps[propKey] = {
      ...columnProps[propKey],
      size
    }
  }
  return columnProps
}

export function assignColProp(key: string, val: any, columnProps: any) {
  const propKey = key !== 'col_props' && key.substr(10)
  const target = !propKey ? columnProps : columnProps[propKey]

  if (!target) {
    columnProps[propKey] = val
  } else if (typeof target === 'object') {
    Object.keys(val).forEach((elemKey) => {
      if (!target[elemKey]) {
        target[elemKey] = val[elemKey]
      }
    })
  } else if (typeof target === 'string' || typeof target === 'number') {
    columnProps[propKey] = {
      ...val,
      size: target
    }
  }
  return columnProps
}

function initialRowStyle(
  { spacing }: ThemeOptions,
  noGutters: boolean
): CSSObject {
  const margin = noGutters ? 0 : `-${spacing}px`
  return {
    display: 'flex',
    flexWrap: 'wrap',
    marginLeft: margin,
    marginRight: margin
  }
}

const noGutterStyle: CSSObject = {
  paddingLeft: 0,
  paddingRight: 0
}
