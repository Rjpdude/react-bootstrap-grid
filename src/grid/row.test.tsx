import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { mount } from 'enzyme'
import { GridTheme } from '../theme/theme'
import { Row } from './row'
import { Col } from './col'

test('Default row', () => {
  const output = renderer
    .create(
      <Row>
        <p>child</p>
      </Row>
    )
    .toJSON()
  expect(output).toMatchSnapshot()
})

test('No gutters', () => {
  const output = renderer
    .create(
      <Row gutters={false}>
        <Col>
          <p>child</p>
        </Col>
      </Row>
    )
    .toJSON()

  expect(output).toMatchSnapshot()
  expect(output).toHaveStyleRule('margin-left', '0')
  expect(output).toHaveStyleRule('margin-right', '0')

  expect(output).toHaveStyleRule('padding-left', '0', {
    modifier: `> ${Col.styledComponent}`
  })
  expect(output).toHaveStyleRule('padding-right', '0', {
    modifier: `> ${Col.styledComponent}`
  })
})

test('Applies column sizes to children', () => {
  const output = mount(
    <Row cols={1} cols_sm={2} cols_md={3} cols_lg={4} cols_xl={5}>
      <Col />
      <Col />
    </Row>
  )
  const children = output.find(Col)

  expect(children.length).toBe(2)

  children.forEach((col) => {
    expect(col.prop('size')).toBe(12)
    expect(col.prop('sm')).toBe(6)
    expect(col.prop('md')).toBe(4)
    expect(col.prop('lg')).toBe(3)
    expect(col.prop('xl')).toBe('20%')
  })
})

test('Applies column sizes to children w/ obj. breakpoint properties', () => {
  const output = mount(
    <Row cols_md={3}>
      <Col md={{ justifyContent: 'center' }} />
      <Col md={{ justifyContent: 'end', size: 9 }} />
    </Row>
  )
  const children = output.find(Col)

  expect(children.at(0).prop('md')).toEqual({
    justifyContent: 'center',
    size: 4
  })

  expect(children.at(1).prop('md')).toEqual({
    justifyContent: 'end',
    size: 9
  })
})

test('Applies column properties', () => {
  const output = mount(
    <Row col_props={{ display: 'flex' }} col_props_xl={{ m: 22 }}>
      <Col />
      <Col display="inline-block" />
    </Row>
  )
  const children = output.find(Col)

  expect(children.at(0).props()).toEqual({
    display: 'flex',
    xl: {
      m: 22
    }
  })

  expect(children.at(1).props()).toEqual({
    display: 'inline-block',
    xl: {
      m: 22
    }
  })
})

test('Applies element style properties', () => {
  const output = renderer
    .create(
      <Row
        p={16}
        justifyContent="between"
        alignSelf="center"
        alignContent="stretch"
      >
        <p>child</p>
      </Row>
    )
    .toJSON()

  expect(output).toMatchSnapshot()
  expect(output).toHaveStyleRule('padding', '16px')
  expect(output).toHaveStyleRule('justify-content', 'space-between')
  expect(output).toHaveStyleRule('align-self', 'center')
  expect(output).toHaveStyleRule('align-content', 'stretch')
})

test('Applies theme context spacing', () => {
  const output = renderer
    .create(
      <GridTheme spacing={8}>
        <Row />
      </GridTheme>
    )
    .toJSON()
  expect(output).toHaveStyleRule('margin-left', '-8px')
  expect(output).toHaveStyleRule('margin-right', '-8px')
})
