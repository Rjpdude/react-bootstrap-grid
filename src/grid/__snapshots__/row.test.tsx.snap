// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Applies element style properties 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: -15px;
  margin-right: -15px;
  padding: 16px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
}

<div
  className="c0 twbs-grid-row"
>
  <p>
    child
  </p>
</div>
`;

exports[`Default row 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: -15px;
  margin-right: -15px;
}

<div
  className="c0 twbs-grid-row"
>
  <p>
    child
  </p>
</div>
`;

exports[`No gutters 1`] = `
.c2 {
  width: 100%;
  position: relative;
  padding-right: 15px;
  padding-left: 15px;
  max-width: 100%;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0;
  margin-right: 0;
}

.c0 > .c1 {
  padding-left: 0;
  padding-right: 0;
}

<div
  className="c0 twbs-grid-row"
>
  <div
    className="c1 c2 twbs-grid-col"
  >
    <p>
      child
    </p>
  </div>
</div>
`;
